sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergUtils.java

```diff
@@ -282,12 +306,24 @@ private static void copyFieldIntoRecord(Record rec, Types.NestedField field, Row
         Optional.ofNullable(value.getDouble(name)).ifPresent(v -> rec.setField(name, v));
         break;
       case DATE:
-        throw new UnsupportedOperationException("Date fields not yet supported");
+        Optional.ofNullable(value.getLogicalTypeValue(name, LocalDate.class))
+            .ifPresent(v -> rec.setField(name, v));
+        break;
       case TIME:
-        throw new UnsupportedOperationException("Time fields not yet supported");
+        Optional.ofNullable(value.getLogicalTypeValue(name, LocalTime.class))
+            .ifPresent(v -> rec.setField(name, v));
+        break;
       case TIMESTAMP:
-        Optional.ofNullable(value.getDateTime(name))
-            .ifPresent(v -> rec.setField(name, v.getMillis()));
+        Object val = value.getValue(name);
+        if (val instanceof Instant) { // case Schema.FieldType.DATETIME
+          rec.setField(name, DateTimeUtil.timestampFromMicros(((Instant) val).getMillis()));
```

Comment:
`DateTimeUtil.timestampFromMicros` expect microseconds, but you're passing milliseconds. If it expects microseconds, you may need to convert milliseconds to microseconds.

Counter-example:

```
import org.apache.iceberg.util.DateTimeUtil;
import org.joda.time.Instant;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.time.LocalDateTime;
import java.time.ZoneId;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;

@RunWith(JUnit4.class)
public class DateTimeUtilsTest {

    @Test
    public void when_DatetimeUtils_TimestampFromMicros_Receives_A_Millis() {
        String timestampStr = "2024-10-07T12:34:56Z";
        Instant instant = Instant.parse(timestampStr);
        LocalDateTime actual = DateTimeUtil.timestampFromMicros(instant.getMillis());

        java.time.Instant javaInstant = java.time.Instant.parse(timestampStr);
        LocalDateTime expected = LocalDateTime.ofInstant(javaInstant, ZoneId.of("UTC"));

        assertThat(actual, equalTo(expected));
    }
}

```

Outcome:
```
java.lang.AssertionError: 
Expected: <2024-10-07T12:34:56>
     but: was <1970-01-21T00:05:04.496>

Reply:
I’ve noticed that the current implementation consistently converts Beam timestamps into java.time.LocalDateTime. However, this approach could lead to issues depending on whether the corresponding Iceberg table column is defined with or without a time zone. Specifically, it may lead to the similar ClassCastException we are currently encountering.

Are you considering adding additional integration tests to address these scenarios?

For your reference, here are a couple of links to the relevant Apache Iceberg / Parquet source code that illustrate the handling of timestamps:

[Iceberg BaseParquetWriter Line 212](https://github.com/apache/iceberg/blob/229d8f6fcd109e6c8943ea7cbb41dab746c6d0ed/parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetWriter.java#L212)

[Iceberg BaseParquetWriter Line 288](https://github.com/apache/iceberg/blob/229d8f6fcd109e6c8943ea7cbb41dab746c6d0ed/parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetWriter.java#L288)

Reply:
You may want to take a look at how Iceberg implemented it for the Spark Runtime. 

[Iceberg Spark Timestamp conversions](https://github.com/apache/iceberg/blob/208ab20dc9ab8bcab3ee525d0ddaba80eeae7609/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java#L138)

Reply:
Great catch! and thanks for all the info and references.

I've been digging for some time and found that Beam has two field types that represent timestamps:
`SqlTypes.DATETIME` -- uses java.time.LocalDateTime 
`FieldType.DATETIME` -- uses org.joda.time.DateTime 

DateTime supports timezone information, but Beam itself doesn't. When this object gets processed in a Beam Row, the timezone information is dropped. Fetching the object will always return a timestamp from UTC.

Because of this, I don't think the connect can support timestamp with TZ yet

Reply:
Actually we might be able to support it by letting the user provide String timestamp representations. We can parse these strings into date time objects then write to Iceberg. Will look into this route soon